### DISCUSSION TITLE
Traits that lib users use but never implement

### DISCUSSION THREAD

**Discussion by tcharding** [rust-bitcoin/rust-bitcoin#1892] [General]
Are there any special API design choices we should be making for traits that are not intended to be implemented ever by users of a library (in `rust-bicoin` and example is `Encodable`)?

One thing I can think of is putting these traits in `prelude::*` so users don't have to import them explicitly. 

Related to that, is there a way to lead users to `prelude`, the compiler is going to warn that `Encodable` is not in scope, its not going to say "just use prelude::*".

**Comment by apoelstra** [comment]
For my part I never use `prelude` modules and think they are a bad idea, but don't mind having one since it's explicitly opt-in and never used by the library itself.

**Comment by Kixunil** [comment]
> Are there any special API design choices we should be making for traits that are not intended to be implemented ever by users of a library 

Yes, sealing them.

> (in `rust-bicoin` and example is `Encodable`)?

That's not actually a good example because I proposed sealing the trait and people argued it's commonly used downstream.

> One thing I can think of is putting these traits in `prelude::*` so users don't have to import them explicitly.

Yeah, this is what many crates do. I agree with Andrew that it's a bad idea but fine to provide as opt-in.

> Related to that, is there a way to lead users to `prelude`, the compiler is going to warn that `Encodable` is not in scope, its not going to say "just use prelude::*".

I don't mind this since I believe `prelude::*` is a bad idea. :)

**Reply by tcharding** [reply]
Just for my understanding; is the reason you guys don't like prelude because it obfuscates where things come from? Or is it some workflow thing (eg grepability)?

**Reply by Kixunil** [reply]
Yes, at least for me.

**Reply by stevenroose** [reply]
I with my vim setup was better with automatically adding imports, because I also prefer not to use preludes, I like that it's visible in your diff when you're using new types in a file.

That said, I would be cautious with marking traits as "users shouldn't implement this". Usually there is no harm in just leaving the traits as is and you can't predict everything users do. F.e. sealing traits means that experimental new bitcoin features might have to fork the library instead of building on them.

(That was why I pushed for `NetworkMessage::Other`, just to name an example; if there are some experimental p2p additions, you want users to be able to defined them themselves in an addon lib and still be able to use all rust-bitcoin-based tooling.)

**Reply by tcharding** [reply]
> you can't predict everything users do

Good point, we should keep this in mind as we "lock down" the lib on the way to v1.0

**Reply by apoelstra** [reply]
True, but conversely, if you seal traits then you can add methods to them in minor version bumps, whereas if they're open you can't. So there's a tradeoff between our own API flexibility and what we let people do.

But I agree we should prefer not to seal things. At least, when we do, we should provide a comment explaining why.

**Reply by stevenroose** [reply]
Curious if you can name the last time you have wanted to add a method to a rust-bitcoin trait :)

**Reply by Kixunil** [reply]
Consensus encoding traits had such changes before.
